Ecr lifecycle

leedm777 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

Add support for [ECR Lifecycle Policies][] to ecs_ecr. Fixes #32003

 [ECR Lifecycle Policies]: https://docs.aws.amazon.com/AmazonECR/latest/userguide/LifecyclePolicies.html

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
`ecs_ecr`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.4.0.0
  config file = /Users/dlee/.ansible.cfg
  configured module search path = [u'/Users/dlee/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.4.0.0/libexec/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

Integration tests provided, and they pass.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
PLAY RECAP *********************************************************************
localhost                  : ok=55   changed=8    unreachable=0    failed=0   
```

leedm777 committed:
Fix copy/pasta for ecs_ecr test names

leedm777 committed:
Add support for lifecycle policies to ecs_ecr

New feature for ecs_ecr to support [ECR Lifecycle Policies][].

Fixes #32003

 [ECR Lifecycle Policies]: https://docs.aws.amazon.com/AmazonECR/latest/userguide/LifecyclePolicies.html

leedm777 committed:
Improve error message for ecs_ecr parsing errors

Replaces the exception and stack trace with a description of what's
actually going wrong from a user perspective.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -235,6 +265,49 @@ def delete_repository_policy(self, registry_id, name):
+
+    def delete_lifecycle_policy(self, registry_id, name):
+        if not self.check_mode:
+            policy = self.ecr.delete_lifecycle_policy(
```
Needs exception handling

willthames commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -235,6 +265,49 @@ def delete_repository_policy(self, registry_id, name):
+
+    def put_lifecycle_policy(self, registry_id, name, policy_text):
+        if not self.check_mode:
+            policy = self.ecr.put_lifecycle_policy(
```
Needs exception handling

willthames commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -235,6 +265,49 @@ def delete_repository_policy(self, registry_id, name):
 
+    def get_lifecycle_policy(self, registry_id, name):
+        try:
+            res = self.ecr.get_lifecycle_policy(
```
Needs exception handling

leedm777 replied:
Exception handling is below at https://github.com/leedm777/ansible/blob/005692ee64d6d775019691180697d756b837d52c/lib/ansible/modules/cloud/amazon/ecs_ecr.py#L437-L443. What else is needed?

willthames replied:
Right, and that wasn't at all obvious to me. I genuinely hadn't noticed the exception handling was there. I prefer the exception handling to be incredibly close to the calling code so that it's obvious. That said, given that this is no worse than currently, I won't block on this.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -263,6 +351,46 @@ def run(ecr, params, verbosity):
+                        original_lifecycle_policy = sort_json_policy_dict(
+                            original_lifecycle_policy)
+
+                    if verbosity >= 3:
```
Please don't change results based on verbosity, this is *very* unexpected.

leedm777 replied:
What's the recommended way to increase details with increased verbosity levels?

willthames replied:
Just return the same stuff for all requests. That's what I expect. 

This and ecs_ecr are the only modules in the entire codebase that return different results based on verbosity. It is not a pattern I would like to see expanded.

Edit: Sorry, this *is* ecs_ecr, making it the only module.

leedm777 replied:
Well, there's lxd_container, lxd_profile, nxos_gir, nxos_gir_profile_management. So I'm not completely alone 😁

Anyways, I take your point. Thinking about it more, what I *really* want here would actually be output when `--diff` is given. But that would be the topic of a different PR.

I'll get this cleaned up. Thanks!

leedm777 replied:
Fixed.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -263,6 +351,46 @@ def run(ecr, params, verbosity):
+                    lifecycle_policy = sort_json_policy_dict(lifecycle_policy)
+                    if verbosity >= 2:
+                        result['lifecycle_policy'] = lifecycle_policy
+                    original_lifecycle_policy = ecr.get_lifecycle_policy(
```
Needs exception handling

willthames commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -263,6 +351,46 @@ def run(ecr, params, verbosity):
+                    result['original_lifecycle_policy'] = original_lifecycle_policy
+
+                if original_lifecycle_policy:
+                    ecr.delete_lifecycle_policy(registry_id, name)
```
Needs exception handling

willthames commented on lib/ansible/modules/cloud/amazon/ecs_ecr.py:
```diff
@@ -43,6 +42,17 @@
+            - JSON or dict that represents the new lifecycle policy
+        required: false
+        version_added: '2.5'
+    delete_lifecycle_policy:
```
Prefer purge_lifecycle_policy to be similar to other modules.

leedm777 replied:
I'll need to also deprecate `delete_policy` and replace it with `purge_policy` for consistency.

leedm777 replied:
Fixed.

willthames reviewed (changes requested):
AWS integration tests that require permissions should include the IAM permissions in a new or existing policy under https://github.com/ansible/ansible/tree/devel/hacking/aws_config/testing_policies - in this case you'll want to create ecs-policy.json

leedm777 committed:
Rename delete policy to purge policy

Marks the `delete_policy` parameter as deprecated, to be removed in
Ansible 2.6.

leedm777 commented:
I've added a `ecs-policy.json` file, but judging from the failed Shippable check I don't think it's working quite right.

willthames commented:
Shippable will need to be updated to use the new policy by a human (for obvious reasons, just adding a file with permissions in it does not change the account permissions) - but the policy allows people to add the permissions to their test account so that they can run it independently of CI too.

cc @mattclay @s-hertel

leedm777 committed:
Add version_added to purge_policy

leedm777 committed:
Remove changing results based on verbosity

What I really want is --diff support, and changing results based on
verbosity is abnormal.

leedm777 commented:
rebased

leedm777 commented:
Anything else you need from me on this PR? I think I addressed all the feedback.

leedm777 commented:
rebased

mattclay commented:
needs_rebase

leedm777 commented:
Rebased.

willthames commented on test/integration/targets/ecs_ecr/tasks/main.yml:
```diff
@@ -248,7 +267,180 @@
+    - name: it should not skip and not change
+      assert:
+        that:
+          - not result|skipped
```
Worth noting #32361 which is why you had to rebase.

Should now be `result is skipped`.

willthames commented:
We've simplified how we manage common AWS parameters for tests.

This is more an FYI than a blocker, as it's so new, but 
https://github.com/willthames/ansible/blob/14b7120d2263b4745f37914a05884809547a6cd6/test/integration/targets/ec2_vpc_route_table/tasks/main.yml#L3-L18

shows how we greatly simplify the amount of boilerplate in AWS test suites.

willthames commented:
@mattclay looks like we need ecr:GetLifecyclePolicy for CI.

s-hertel committed:
Ensure repository name is lowercase

s-hertel committed:
Fix deprecation cycle to 4 releases

s-hertel committed:
Use a YAML anchor for credentials

s-hertel committed:
Remove filters from assertions

s-hertel committed:
Add minimal permissions needed

alanroche commented:
Hi, 
Any progress on this PR? 
This would be a great feature to have, - we need to resort to using the CLI for updating life-cycle polices outside of the ecs_ecr module as a workaround. 
Thanks...

breyer commented:
We need the module, too.

agustingarciaflores commented:
Hi! Any new progress on this feature? Seems like it would be incredibly useful for some folks, including myself ;)

gundalow commented:
If people would like to help we need
* Rebase
* Revert CHANGELOG.md

People to test and report if it works

SpamapS commented:
https://github.com/ansible/ansible/pull/48997 has all of that done, just a couple of lint failures.

SpamapS commented:
Lint errors fixed. Please do go review that PR (#48997) if you want this feature.

gundalow commented:
@SpamapS that's brilliant, thank you.
Closing this PR please review #48997
